### PR TITLE
Add basic CI for smoke test and pre-commit

### PR DESCRIPTION
**Pull request by unkcpz**
~~The proposed format for the pdos input data and schema check for the file.~~
~~The schema for the PDOS is TBD.~~

EDITED: the schema part of this implementation is redirect to PR #18. This PR will then change to having only address the CI and pre-commit.

### CODE REVIEW

**Comment by unkcpz** [comment]
Hi @csadorf, can you have a look at this PR, mostly the schema added and adding pre-commit to clean up the code.

**Comment by unkcpz** [comment]
@dou-du the voila-test is failed, I copy and paste from the origin test which is not triggered before. Can you have a look and see how it can be fixed? Thanks.

**Comment by csadorf** [comment]
> Hi @csadorf, can you have a look at this PR, mostly the schema added and adding pre-commit to clean up the code.

Yes, of course, but maybe we can split the schema from the changes related to CI and testing into a separate PR? Just to be able to keep the discussion focused.

**Comment by unkcpz** [comment]
> Yes, of course, but maybe we can split the schema from the changes related to CI and testing into a separate PR? Just to be able to keep the discussion focused.

Sure, will do that.

File: widget_bandsplot/schemas/pdos.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+    "$id": "https://example.com/pdos",
```
Comment (csadorf) [inline widget_bandsplot/schemas/pdos.json:2]: You could use the GitHub org as a namespace here, similar to [here](https://github.com/aiidalab/aiidalab/blob/37c92fbeee0a9c02b8211e1c1767bb849b84eb8f/aiidalab/registry/schemas/app.schema.json#L3).

File: widget_bandsplot/schemas/pdos.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "description": "JSON Schema for an PDOS entry",
```
Comment (csadorf) [inline widget_bandsplot/schemas/pdos.json:4]:
```suggestion
    "description": "JSON Schema for a PDOS entry",
```

File: widget_bandsplot/schemas/pdos.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (csadorf) [inline widget_bandsplot/schemas/pdos.json:1]: Is this file referenced anywhere in the code? You could use python-jsonschema to validate input against this schema.

File: widget_bandsplot/schemas/pdos.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+                        "type": "array",
+                        "items": {"type": "number"}
+                    },
+                    "y": {
+                        "type": "array",
+                        "items": {"type": "number"}
+                    },
+                    "lineStyle": {"enum": ["solid", "dash"]},
+                    "borderColor": {"type": "string"},
+                    "backgroundColor": {"type": "string"}
```
Comment (csadorf) [inline widget_bandsplot/schemas/pdos.json:26]: Should we maybe place all representation-related (i.e. not data) entries in a sub-property?

**Comment by unkcpz** [comment]
Hi @csadorf, I move the schema part to https://github.com/osscar-org/widget-bandsplot/pull/18. But seems you have already started reviewing, do I need to reply and update this PR?

**Comment by csadorf** [comment]
> Hi @csadorf, I move the schema part to #18. But seems you have already started reviewing, do I need to reply and update this PR?

Just address my comments on the new PR and we can continue the discussion there.

File: widget_bandsplot/schemas/pdos.json (hunk 1 above)
Comment (unkcpz) [reply widget_bandsplot/schemas/pdos.json:2]: But the tricky part is we don't have a version released with this file, so I'll use the file URL of the branch. But I can open an issue and change this after a new version with this released.

File: widget_bandsplot/schemas/pdos.json (hunk 3 above)
Comment (unkcpz) [reply widget_bandsplot/schemas/pdos.json:1]: Yes, I use it in the test and for the validation of the example JSON file by pre-commit.

File: widget_bandsplot/schemas/pdos.json (hunk 4 above)
Comment (unkcpz) [reply widget_bandsplot/schemas/pdos.json:26]: I think this format reflects what matplotlib did for plotting lines. and it is simpler at the same level for all parameters.

**Comment by unkcpz** [comment]
@csadorf thanks! The PR is redirected to https://github.com/osscar-org/widget-bandsplot/pull/18 with your suggestion addressed.

**Comment by yakutovicha** [comment]
Just to clarify: my input is needed for the #18 and not here, correct @unkcpz?

File: widget_bandsplot/schemas/pdos.json (hunk 1 above)
Comment (csadorf) [reply widget_bandsplot/schemas/pdos.json:2]: You can interpret this more as a name than an actual URL. Just use the URL it will have after merge.

**Comment by csadorf** [comment]
> @csadorf thanks! The PR is redirected to #18 with your suggestion addressed.

I'm a bit confused, it seems that #18 still has a lot of changes that are not directly related to the schema?

File: widget_bandsplot/schemas/pdos.json (hunk 3 above)
Comment (csadorf) [reply widget_bandsplot/schemas/pdos.json:1]: What about my suggestion to validate the input during runtime?

File: widget_bandsplot/schemas/pdos.json (hunk 4 above)
Comment (csadorf) [reply widget_bandsplot/schemas/pdos.json:26]: Sure.

**Comment by unkcpz** [comment]
> I'm a bit confused, it seems that https://github.com/osscar-org/widget-bandsplot/pull/18 still has a lot of changes that are not directly related to the schema?

Which file? The CI and unit tests are for schema test and folder `example` is renamed to `examples`.

File: widget_bandsplot/schemas/pdos.json (hunk 3 above)
Comment (unkcpz) [reply widget_bandsplot/schemas/pdos.json:1]: I see, good point! I add the check when load the dos inputs in `BandsWidget` Class, please have a look at the latest commit https://github.com/osscar-org/widget-bandsplot/pull/18/commits/0104b8cb44e585a2eaedd32a288a3c99c730be15

Comment (csadorf) [reply widget_bandsplot/schemas/pdos.json:1]: Great, thanks!

**Comment by csadorf** [comment]
> > I'm a bit confused, it seems that #18 still has a lot of changes that are not directly related to the schema?
> 
> Which file? The CI and unit tests are for schema test and folder `example` is renamed to `examples`.

You're right, I didn't realize the tests were strictly related to the schema. All good.

**Comment by dou-du** [comment]
> @dou-du the voila-test is failed, I copy and paste from the origin test which is not triggered before. Can you have a look and see how it can be fixed? Thanks.

I just fixed it. It is done!

**Pull request merged**